File: app/src/main/java/net/tlalka/android/fiszki/fragments/LanguageDialogFragment.java
package net.tlalka.android.fiszki.fragments;

import android.app.AlertDialog;
import android.app.Dialog;
import android.app.DialogFragment;
import android.content.Context;
import android.content.DialogInterface;
import android.os.Bundle;
import net.tlalka.android.fiszki.R;
import net.tlalka.android.fiszki.adapters.LanguageAdapter;
import net.tlalka.android.fiszki.models.dto.LanguagesDto;
import net.tlalka.android.fiszki.models.types.LanguageType;
import org.parceler.Parcels;

import java.util.List;

public class LanguageDialogFragment extends DialogFragment {

    private DialogListener listener;
    private LanguagesDto languagesDto;

    public interface DialogListener {
        void onLanguageSelected(LanguageType languageType);
    }

    private class LanguageListener implements DialogInterface.OnClickListener {

        @Override
        public void onClick(DialogInterface dialog, int which) {
            listener.onLanguageSelected(languagesDto.getLanguages().get(which));
        }
    }

    public static LanguageDialogFragment getInstance(List<LanguageType> languageTypes) {
        Bundle bundle = new Bundle();
        bundle.putParcelable(LanguagesDto.class.getName(), Parcels.wrap(new LanguagesDto(languageTypes)));

        LanguageDialogFragment fragment = new LanguageDialogFragment();
        fragment.setArguments(bundle);
        return fragment;
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        this.listener = (DialogListener) context;
    }

    @Override
    public Dialog onCreateDialog(Bundle bundle) {
        languagesDto = Parcels.unwrap(getArguments().getParcelable(LanguagesDto.class.getName()));

        return new AlertDialog.Builder(getActivity())
                .setIcon(R.drawable.icon_lesson_stats)
                .setTitle(R.string.language_dialog_title)
                .setAdapter(new LanguageAdapter(getActivity(), languagesDto.getLanguages()), new LanguageListener())
                .create();
    }
}

File: app/src/main/kotlin/net/tlalka/fiszki/view/fragments/LanguageDialogFragment.kt
package net.tlalka.fiszki.view.fragments

import android.app.Activity
import android.app.AlertDialog
import android.app.Dialog
import android.app.DialogFragment
import android.content.DialogInterface
import android.os.Bundle
import net.tlalka.fiszki.R
import net.tlalka.fiszki.model.dto.parcel.LanguagesDto
import net.tlalka.fiszki.model.types.LanguageType
import net.tlalka.fiszki.view.adapters.LanguageAdapter

class LanguageDialogFragment : DialogFragment() {

    private var listener: DialogListener? = null
    private var languagesDto: LanguagesDto? = null

    interface DialogListener {
        fun onLanguageSelected(languageType: LanguageType)
    }

    private inner class LanguageListener : DialogInterface.OnClickListener {
        override fun onClick(dialog: DialogInterface, which: Int) {
            listener?.onLanguageSelected(languagesDto!!.languages[which])
        }
    }

    override fun onAttach(activity: Activity) {
        super.onAttach(activity)
        listener = activity as DialogListener
    }

    override fun onCreateDialog(bundle: Bundle): Dialog {
        languagesDto = arguments.getParcelable(LanguagesDto::class.java.name)

        return AlertDialog.Builder(activity)
                .setIcon(R.drawable.language_icon)
                .setTitle(R.string.language_dialog_title)
                .setAdapter(LanguageAdapter(activity, languagesDto!!.languages), LanguageListener())
                .create()
    }

    companion object {
        fun getInstance(languageTypes: List<LanguageType>): LanguageDialogFragment {
            val bundle = Bundle()
            bundle.putParcelable(LanguagesDto::class.java.name, LanguagesDto(languageTypes))

            val fragment = LanguageDialogFragment()
            fragment.arguments = bundle
            return fragment
        }
    }
}
